doc/man-source/xmode2.1
.TH MODE2 "1" "Last change: Aug 2015" "mode2 @version@" "User Commands"
.SH NAME
xmode2 - Show the pulse/space length of infrared signals
.SH SYNOPSIS
.P
\fBxmode2\fR [\fIoptions\fR]

.SH DESCRIPTION
.P
The main purpose of \fBxmode2\fR is to check operation of your
home-brew LIRC receiver hardware and to see the IR waveform of the
remote controller without an expensive oscilloscope. Very useful for
debugging. Of course this program won't work with hardware that decodes
the signals itself like e.g. TV cards or the Irman i. e., it only
works with LIRC_MODE_MODE2 type drivers.
.P
\fBxmode2\fR prints a simple graphics display of the pulse/space lengths. It
accepts a few commands including time base changes; see the xmode2 -h output.
.P
lirc-lsplugins(1) allows checking if a driver is a LIRC_MODE2 type one.


.SH OPTIONS

.TP
\fB\-h\fR \fB\-\-help\fR
display usage summary
.TP
\fB\-v\fR \fB\-\-version\fR
display version
.TP
\fB\-d\fR \fB\-\-device\fR=\fIdevice\fR
read from given device
.TP
\fB\-H\fR \fB\-\-driver\fR=\fIdriver\fR
use given driver
.TP
\fB\-m\fR \fB\-\-mode\fR
enable alternative display mode
.TP
\fB\-r\fR \fB\-\-raw\fR
access device directly

.SH RUNNING AS ROOT
In many cases xmode2 needs to run as root to access devices not available
to regular users. On the other hand, running as root creates problems
such as log files owned by root, security concerns etc.
.P
In order to cope with this, xmode2 by default drops root privileges
after opening the input device. This support is based on that root
permissions are acquired using sudo(1) e. g., using
.nf

        $ sudo xmode2 --driver default --device /dev/lirc0

.fi
If not using sudo, the same behaviour could be accomplished using the
SUDO_USER environment variable e. g.,
.nf

        # SUDO_USER=$LOGNAME xmode2 --raw /dev/lirc0

.fi
The --keep-root option will make xmode2 keep it's root privileges for the
complete run.


.SH "DRIVER LOADING"
Drivers are loaded dynamically. The search path used for this is determined
by (falling priority):
.IP \- 2
The --plugindir option.
.IP \- 2
The 'plugindir' entry in  the [lircd] section of the lirc_options.conf file.
.IP \- 2
The environment variable LIRC_PLUGINDIR.
.IP \- 2
A hardcoded default (usually /usr/lib[64]/lirc/plugins).

.SH FILES

.TP 4
.B ~/.cache/xmode2.log
Debug output. Setting the XDG_CACHE_HOME environment variable relocates this
file to $XDG_CACHE_HOME/mode2.log

.SH "SEE ALSO"
mode2(1)
.br
lirc-lsplugins(1)

The documentation for
.B lirc
is maintained as html pages. They are located under html/ in the
documentation directory.


NEWS
0.9.4-devel
  * Issues: https://sf.net/p/lirc/tickets/milestone/0.9.4

  * The caraca driver has been retired (still available in attic/).
  * The iguanair driver has been moved out of tree and is now
    maintained by Iguanaworks, Inc. See drivers/iguanair/README.
  * The irman driver has been moved out of tree. From 0.5.1 the
    libirman package contains the driver. See drivers/irman/README
  * The yard2 driver has been moved out of tree and is now maintained
    by the yard2 project. See driver/yard2/README.
  * New upstream debian packaging with which Debian users can create
    .deb packages. Ubuntu packages are also available in PPAs. See
     debian/INSTALL (#137).
  * New driver girs for the GIRS protocol, gives LIRC Arduino support.
    See new info in driver documentation.
  * Support for non-modulated signals added (#132).
  * As shipped, lircd now uses the devinput driver and the new 'auto'
    device, should work for all remotes supported by kernel (#106).
  * Enhanced lircd-setup tool (#123, partly)
    - Integrated mode2 and irw tests of current configuration.
    - Documentation links.
    - More consistent data based on new database.
    - Multiple bugfixes.
  * Several patches to make mode2 work with all drivers(#123, partly)
  * Configuration database overhaul (#123, partly)
    - Database lives as yaml files in configs/.
    - Data is collected from drivers, remotes and static files.
    - Better ways for drivers to enumerate devices.
    - There is an unstable python database module used by
      all tools.
  * devinput.lircd.conf is generated at build time to match host kernel,
    new script lirc-make-devinput shipped to re-generate it (#153)
  * A new script lircd-setup runs as a one-shot service before lircd,
    performing setup commands as root (#121).
  * Several new examples in contrib/, notably lircrc templates for
    many popular LIRC applications and improved udev rules.
  * The default driver does not require root perms if the protocol
    is correct (#133).
  * Add an "auto" device to devinput driver for single-device
    usecase (#134).
  * irrecord has been taught to handle remotes without gap (#98).
  * The --uinput lircd option is deprecated, obsoleted by the
    lircd.uinput.service systemd service. See lircd-uinput(8).
  * CLI options for mode2, xmode2 and irrecord has been streamlined.
    All tools falls back to lircd defaults where it makes sense.

  * The driver API has been expanded with functions to enumerate
    devices. See configs/README and lib/driver.h.
  * New logging API with hooks for enhanced messages and filtering
    (also #129).
  * The out-of-tree build interface is enhanced with hooks for external
    drivers to rebuild documentation and add a config file.

  * lircd: Fix too-short gap  when sending/blasting (#158).
  * irrecord fixes:
    - Buggy handling of not decoded buttons in irrecord (#146).
    - Repeated buttons problems (#149)
    - Endless loop when key is not recognized (#146)
    - Endless loop on idendifying toggle mark (#139)
  * Remotes become too sensitive in 0.9.3 (#148)
  * Fix --plugindir not being respected (#144).
  * Fix crash for non-existing device (#145).
  * Fix gid and supplementary groups in CLI apps dropping root.
  * Fix a crash in irrecord when invoked without options.
  * Fix bad parsing of --options-file unless given first (#165).
  * Don't require write access if not needed in default.c (#133).
  * Clean up docs, remove circular links (#136). Documentation is
    now only in lirc.org/html.

  * Upstream selinux updated to match current lircd requirements.
    See: https://github.com/fedora-selinux/selinux-policy/pull/41
  * The lirc.4 manpage (/dev/lirc* interface) is upstreamed to the
    linux kernel manpages project (#

0.9.3  4/09/15
  * Digest of 0.9.3pre1..0.9.3pre3

  * New build dependency: xsltproc (sometimes coming with libxml).
  * The irexec semantics have been changed so that the invoked
    programs run disconnected from the main irexec process.
    See irexec(1) (#92).
  * irrecord has got some love:
     + It handles multiple definitions of same symbol OK (#78)
     + Warns about single-key config files (#77)
     + Checks for ambient light disturbances (#76)
     + New option allows updating of existing template file.
     + Template handling bug fixed
     + -l --list-namespace option parsing bug fixed (#101).
     + Minor fixes to irrecord-ui including -k option parsing fixed.
  * mode2 has been updated
     +  handle also LIRCCODE devices (#94),
     +  Drop root privileges (#91)
     +  Add  new scope-like display
     +  Handle the--driver-options switch
  * xmode2 has been updated to handle udp driver, the new
    --driver-options switch + some general overhaul.
  * irrecord has been split into a library and a cli interface
    (#36). The library interface is unstable.

  * The select() calls in lircd, lircrcd and library routines have
    been replaced with poll() to avoid crashes when file descriptors
    becomes > 1024 (#109).
  * Set supplementary groups using effective-user (RH bz 1253907).
  * Work around kernel bug in /dev/lirc0 output (#119).
  * lircmd: Several option parsing bugs fixed, one of which
    blocked the systemd service (#104, RH bz 1234155).
  * lirc-lsremotes now displays logs after parsing files, making
    more parsing warnings visible.
  * lirc-config-tool is fixed to comply with guidelines (#105).
  * Bug in irexec when using modes fixed (#102).
  * Create log directory if required (#124).
  * lirc-setup: Fixed name clash on recent python versions.

  * irtoy: New support for debug LEDs. New documentation
    file, support in lirc-setup and driver table doc added.
  * The udp driver is patched to allow adjustable clockrates,
    see new udp_driver.html driver doc.
  * The mplay driver has been updated with patches which has
    been available "out there".

  * The library routines for parsing lircd.conf now warns for

tools/xmode2.cpp
/***************************************************************************
** xmode2.c ****************************************************************
****************************************************************************
*
* xmode2 - shows the ir waveform of an IR signal
*
* patched together on Feb. 18th 1999 by
* Heinrich Langos <heinrich@mad.scientist.com>
*
* This program is based on the smode2.c file by Sinkovics Zoltan
* <sinko@szarvas.hu> which is a part of the LIRC distribution. It is
* just a conversion from svga to X with some basic support for resizing.
* I copied most of this comment.
*
* This program is based on the mode2.c file which is a part of the
* LIRC distribution. The main purpose of this program is to check
* operation of LIRC receiver hardware, and to see the IR waveform of
* the remote controller without an expensive oscilloscope. The time
* division is variable from 1 ms/div to extremely high values (integer
* type) but there is no point increasing this value above 20 ms/div,
* because one pulse is about 1 ms. I think this kind of presentation
* is much more exciting than the simple pulse&space output showed by
* mode2.
*
* Usage: xmode2 [-t (ms/div)] , default division is 5 ms/div
*
*
* compile: gcc -o xmode2 xmode2.c -L/usr/X11R6/lib -lX11
*
* version 0.01  Feb 18 1999
*   initial release
*
* version 0.02  Aug 24 1999
*   using select() to make the whole thing more responsive
* */

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdarg.h>
#include <unistd.h>
#include <fcntl.h>
#include <getopt.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/ioctl.h>
#include <limits.h>
#include <errno.h>
#include <syslog.h>
#include <time.h>

#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/cursorfont.h>

#include "lirc_private.h"

const char* const  MSG_USE_RAW =
	"Please use the --raw option to access the device directly instead.";

const char* const MSG_BAD_DRIVER =
	"This program does not work for this driver and hardware.\n"
	"(the driver is not a LIRC_MODE2 type driver)\n"
	" You can test this receiver using mode2(1).";

const char* const MSG_MODE2_DRIVER =
	"This program is only intended for receivers supporting the\n"
	" pulse/space layer. Note that this is no error, but this program\n"
	" simply makes no sense for your receiver.\n"
	" You can test this receiver using mode2(1).";

static const logchannel_t logchannel = LOG_APP;

Display* d1;
Window w0, w1;			/*w0 = root */
char w1_wname[] = "xmode2";
char w1_iname[] = "xmode2";
char font1_name[] = "-misc-fixed-*-r-*-*-12-*-*-*-*-*-iso8859-1";

int w1_x = 0, w1_y = 0;
unsigned int w1_w = 640, w1_h = 480, w1_border = 0;

XFontStruct* f1_str;
XColor xc1, xc2;
Colormap cm1;
XGCValues gcval1;
GC gc1, gc2;
XSetWindowAttributes winatt1;

long event_mask1;
XEvent event_return1;

static int point = 0;
static int showText = 0;
static int div_ = 5;
static int dmode = 0;
static struct stat s;
static int use_stdin = 0;
static int use_raw_access = 0;

static const char* device = "";
static char* geometry = NULL;

static struct option options[] = {
	{"help",           no_argument,	      NULL, 'h'},
	{"version",        no_argument,	      NULL, 'v'},
	{"debug",          required_argument, NULL, 'D'},
	{"device",         required_argument, NULL, 'd'},
	{"driver",         required_argument, NULL, 'H'},
	{"keep-root",      required_argument, NULL, 'k'},
	{"geometry",       required_argument, NULL, 'g'},
	{"timediv",        required_argument, NULL, 't'},
	{"mode",           no_argument,       NULL, 'm'},
	{"raw",            no_argument,       NULL, 'r'},
	{"plugindir",      required_argument, NULL, 'U'},
	{"driver-options", required_argument, NULL, 'A'},
	{0, 0, 0, 0}
};


static const char* const help =
	"\nUsage: xmode2 [options]\n"
	"\nOptions:\n"
	"    -d --device=device\t\tRead from given device\n"
	"    -H --driver=driver\t\tUse given driver\n"
	"    -U --plugindir=dir\t\tLoad drivers from given dir\n"
	"    -g --geometry=geometry\twindow geometry\n"
	"    -t --timediv=value\t\tms per unit\n"
	"    -m --mode\t\t\tEnable alternative display mode\n"
	"    -k --keep-root\t\tKeep root privileges\n"
	"    -r --raw\t\t\tAccess device directly\n"
	"    -O --options-file\t\tUse alternative lirc_options.conf file\n"
	"    -A --driver-options=key:value[|key:value...]\n"
	"\t\t\t\tSet driver options\n"
	"    -D --loglevel=level\t\t'error', 'info', 'notice',... or 3..10\n"
	"    -h --help\t\t\tDisplay usage summary\n"
	"    -v --version\t\tDisplay version\n\n"
	"The window responds to the following keys:\n"
	"    .1, .2, .5, 1, 2 & 5 Set the timebase (ms)\n"
	"    m to Toggle the display mode\n"
	"    q to Quit\n";


static void add_defaults(void)
{
	const char* const device = options_getstring("lircd:device");
	const char* const driver = options_getstring("lircd:driver");
	const char* const lircd_level = options_getstring("lircd:debug");
	const char* const plugindir = options_getstring("lircd:plugindir");
	char level[12];

	if (lircd_level != NULL) {
		strncpy(level, lircd_level, sizeof(level) - 1);
	} else {
		snprintf(level, sizeof(level), " %d", lirc_log_defaultlevel());
	}
	const char* const defaults[] = {
		"xmode2:plugindir",  	plugindir ? plugindir : PLUGINDIR,
		"xmode2:debug",		level,
		"xmode2:device",     	device ? device : LIRC_DRIVER_DEVICE,
		"xmode2:driver",        driver ? driver : "default",
		"xmode2:analyse",	"False",
		"xmode2:force",		"False",
		"xmode2:dynamic-codes",	"False",
		"xmode2:list_namespace","False",
		"xmode2:disable-namespace", "False",
		(const char*)NULL,	    (const char*)NULL
	};
	options_add_defaults(defaults);
}


static void parse_options(int argc, char** const argv)
{
	int c;

	add_defaults();
	char driver[64];
	const char* const optstring =  "g:hvdD:U:H:t:mrA:";

	strcpy(driver, "default");
	while ((c = getopt_long(argc, argv, optstring, options, NULL)) != -1) {
		switch (c) {
		case 'D':
			if (string2loglevel(optarg) == LIRC_BADLEVEL) {
				fprintf(stderr, "Bad debug level: %s\n", optarg);
				exit(EXIT_FAILURE);
			}
			options_set_opt("xmode2:debug", optarg);
			break;
		case 'h':
			puts(help);
			exit(EXIT_SUCCESS);
		case 'H':
			options_set_opt("xmode2:driver", optarg);
			break;
		case 'v':
			printf("%s %s\n", progname, VERSION);
			exit(EXIT_SUCCESS);
		case 'd':
			options_set_opt("xmode2:device", optarg);
			break;
		case 'g':
			geometry = optarg;
			break;
		case 't':       /* timediv */
			div_ = strtol(optarg, NULL, 10);
			break;
		case 'k':
			unsetenv("SUDO_USER");
			break;
		case 'm':
			dmode = 1;
			break;
		case 'r':
			use_raw_access = 1;
			break;
		case 'U':
			options_set_opt("xmode2:plugindir", optarg);
			break;
		case 'A':
			options_set_opt("lircd:driver-options", optarg);
			break;
		default:
			printf("Usage: %s [options]\n", progname);
			exit(EXIT_FAILURE);
		}
	}
	if (optind < argc) {
		fprintf(stderr, "%s: too many arguments\n", progname);
		exit(EXIT_FAILURE);
	}
	options_set_opt("lircd:plugindir",
			options_getstring("xmode2:plugindir"));
	if (hw_choose_driver(driver) != 0) {
		fprintf(stderr, "Driver `%s' not found", driver);
		fputs(" (wrong or missing -U/--plugindir?)\n", stderr);
		hw_print_drivers(stderr);
		exit(EXIT_FAILURE);
	}
	strncpy(driver,
		options_getstring("xmode2:driver"),
		sizeof(driver) - 1);
}


void initscreen(char* geometry)
{
	d1 = XOpenDisplay(0);
	if (d1 == NULL) {
		puts("Can't open display.");
		exit(0);
	}

	if (geometry != NULL)
		XParseGeometry(geometry, &w1_x, &w1_y, &w1_w, &w1_h);

	/*Aufbau der XWindowsAttribStr */
	w0 = DefaultRootWindow(d1);
	winatt1.background_pixel = BlackPixel(d1, 0);
	winatt1.backing_store = WhenMapped;
	winatt1.event_mask = KeyPressMask | StructureNotifyMask | ExposureMask;
	w1 = XCreateWindow(d1, w0, w1_x, w1_y, w1_w, w1_h, w1_border, CopyFromParent, InputOutput, CopyFromParent,
			   CWBackPixel | CWBackingStore | CWEventMask, &winatt1);

	XStoreName(d1, w1, w1_wname);
	XSetIconName(d1, w1, w1_iname);
	XMapWindow(d1, w1);

	cm1 = DefaultColormap(d1, 0);
	if (!XAllocNamedColor(d1, cm1, "blue", &xc1, &xc2))
		puts("couldn't allocate blue color");
	f1_str = XLoadQueryFont(d1, font1_name);
	if (f1_str == NULL) {
		puts("could't load font");
		exit(EXIT_FAILURE);
	}

	gcval1.foreground = xc1.pixel;
	gcval1.font = f1_str->fid;
	gcval1.line_style = LineSolid;
	gc1 = XCreateGC(d1, w1, GCForeground | GCLineStyle, &gcval1);
	gcval1.foreground = WhitePixel(d1, 0);
	gc2 = XCreateGC(d1, w1, GCForeground | GCLineStyle | GCFont, &gcval1);
}


void closescreen(void)
{
	XUnmapWindow(d1, w1);
	XCloseDisplay(d1);
}

void drawGrid(int div)
{
	char textbuffer[80];
	int x;

	XClearWindow(d1, w1);
	for (x = 0; x < (int)w1_w; x += 10)
		XDrawLine(d1, w1, gc1,  x, 0,  x, w1_h);

	sprintf(textbuffer, "%5.3f ms/div", div/100.0);
	XDrawString(d1, w1, gc2, w1_w - 100, 10, textbuffer, strlen(textbuffer));
	XFlush(d1);
}

int main(int argc, char** argv)
{
	fd_set rfds;
	int xfd, maxfd;

	int fd;
	__u32 mode;
	lirc_t data = 0;
	lirc_t x1, y1, dx;
	int result;
	char textbuffer[80];
	const char* opt;
	char logpath[256];
	const loglevel_t level = options_get_app_loglevel("xmode2");

	hw_choose_driver(NULL);
	options_load(argc, argv, NULL, parse_options);

	lirc_log_get_clientlog("xmode2", logpath, sizeof(logpath));
	(void)unlink(logpath);
	lirc_log_set_file(logpath);
	lirc_log_open("xmode2", 0, level);

	device = options_getstring("xmode2:device");
	if (strcmp(device, LIRCD) == 0) {
		fputs("Refusing to connect to lircd socket\n", stderr);
		return EXIT_FAILURE;
	}
	if (use_raw_access) {
		printf("Using raw access on device %s\n", device);
	} else {
		printf("Using driver %s on device %s\n",
		       options_getstring("xmode2:driver"),
		       device);
	}
	if (!isatty(STDIN_FILENO)) {
		use_stdin = 1;
		fd = STDIN_FILENO;
	} else if (use_raw_access) {
		fd = open(device, O_RDONLY);
		if (fd == -1) {
			perrorf("Error opening %s", device ? device : "Null");
			exit(EXIT_FAILURE);
		}
		;

		if ((fstat(fd, &s) != -1) && (S_ISFIFO(s.st_mode))) {
			/* can't do ioctls on a pipe */
		} else if ((fstat(fd, &s) != -1) && (!S_ISCHR(s.st_mode))) {
			fprintf(stderr,
				"%s is not a character device\n", device);
			fputs("Use the -d option to specify the correct device\n",
			      stderr);
			close(fd);
			exit(EXIT_FAILURE);
		} else if (ioctl(fd, LIRC_GET_REC_MODE, &mode) == -1) {
			fputs(MSG_MODE2_DRIVER, stderr);
			exit(EXIT_FAILURE);
		}
	} else {
		curr_driver->open_func(device);
		opt = options_getstring("lircd:driver-options");
		if (opt != NULL)
			if (drv_handle_options(opt) != 0) {
				fprintf(stderr,
				"Cannot set driver (%s) options (%s)\n",
				curr_driver->name, opt);
				return EXIT_FAILURE;
			}
		if (curr_driver->init_func  && !curr_driver->init_func()) {
			fputs("Cannot initialize hardware\n", stderr);
			exit(EXIT_FAILURE);
		}

		fd = curr_driver->fd;   /* please compiler */
		mode = curr_driver->rec_mode;
		if (mode != LIRC_MODE_MODE2) {
			if (strcmp(curr_driver->name, "default") == 0) {
				puts(MSG_USE_RAW);
			} else {
				puts(MSG_BAD_DRIVER);
			}
			exit(EXIT_FAILURE);
		}
	}

	if (geteuid() == 0)
		drop_root_cli(setuid);
	initscreen(geometry);
	xfd = XConnectionNumber(d1);
	maxfd = fd > xfd ? fd : xfd;
	y1 = 20;
	x1 = 0;
	drawGrid(div_);
	while (1) {
		while (XPending(d1) > 0) {
			XNextEvent(d1, &event_return1);
			switch (event_return1.type) {
			case KeyPress:
				if (1 == XLookupString(&event_return1.xkey,
						       textbuffer,
						       sizeof(textbuffer),
						       NULL, NULL)) {
					switch (textbuffer[0]) {
					case 'q':
					closescreen();
					exit(1);
					case 'm':
						dmode = !dmode;
						break;
					/*
					 * Switch the time-base, the screen is not cleared because it is useful to see
					 * the IR data with different time-bases on the same screen. The new time-base
					 * text is drawn to the screen and the waveform is set to start on a new line.
					 */
					case '.':
						point = 1;	/* Remember the decimal point */
						break;
					case '1':
						div_ = 100;	/* 1ms, 1000us per division / 10 pixels per division. */
						showText = 1;
						break;
					case '2':
						div_ = 200;
						showText = 1;
						break;
					case '5':
						div_ = 500;
						showText = 1;
						break;
					}
					if (showText) {
						showText = 0;
						if (point) {
							point = 0;
							div_ /= 10;
						}
						y1 += 25;
						sprintf(textbuffer, "%5.3f ms/div", div_ / 100.0);
						XDrawString(d1, w1, gc2,
							    w1_w - 100,
							    y1,
							    textbuffer,
							    strlen(textbuffer));
						y1 += 5;
						x1 = 0;
					}
				}
				break;
			case Expose:
			case ConfigureNotify:
				switch (event_return1.type) {
				case Expose:
					break;
				case ConfigureNotify:
					if ((int) w1_w == event_return1.xconfigure.width &&
					    (int) w1_h == event_return1.xconfigure.height)
						continue;

					w1_w = event_return1.xconfigure.width;
					w1_h = event_return1.xconfigure.height;
					break;
				}
				y1 = 20;
				x1 = 0;
				drawGrid(div_);
				break;
			default:
				break;
			}
		}

		FD_ZERO(&rfds);
		FD_SET(fd, &rfds);
		FD_SET(xfd, &rfds);

		select(maxfd + 1, &rfds, NULL, NULL, NULL);

		if (FD_ISSET(fd, &rfds)) {
			if (use_stdin) {
				static int space = 1;
				unsigned long scan;

				if (space)
					result = fscanf(stdin, "space %ld\n", &scan);
				else
					result = fscanf(stdin, "pulse %ld\n", &scan);
				if (result == 1) {
					data = (lirc_t)scan;
					if (!space)
						data |= PULSE_BIT;
				} else {
					fd = STDOUT_FILENO;
				}
				space = !space;
			} else {
				/*
				 * Must use the driver read function, the UDP driver reformats the data!
				 */
				data = curr_driver->readdata(0);
				if (data == 0) {
					fprintf(stderr, "readdata() failed\n");
					result = 0;
				} else {
					result = 1;
				}
			}
			if (result != 0) {
				if lirc_log_is_enabled_for(LIRC_DEBUG) {
					if (data & PULSE_BIT)
						printf("%.8x\t", data);
					else
						printf("%.8x\n", data);
				}
				dx = (data & PULSE_MASK) / (div_);
				if (dx > 400) {
					if (!dmode)
						y1 += 15;
					else
						y1++;
					x1 = 0;
				} else {
					if (x1 == 0) {
						if (!dmode)
							XDrawLine(d1, w1, gc2, x1, y1 + 10, x1 + 10, y1 + 10);
						x1 += 10;
						if (!dmode)
							XDrawLine(d1, w1, gc2, x1, y1 + 10, x1, y1);
					}
					if (x1 < (int) w1_w) {
						if (dmode) {
							if (data & PULSE_BIT)
								XDrawLine(d1, w1, gc2, x1, y1, x1 + dx, y1);
							x1 += dx;
						} else {
							XDrawLine(d1, w1, gc2,
								x1,      ((data & PULSE_BIT) ? y1 : y1 + 10),
								x1 + dx, ((data & PULSE_BIT) ? y1 : y1 + 10));
							x1 += dx;
							XDrawLine(d1, w1, gc2,
								x1, ((data & PULSE_BIT) ? y1 : y1 + 10),
								x1, ((data & PULSE_BIT) ? y1 + 10 : y1));
						}
					}
				}
				if (y1 > (lirc_t) w1_h) {
					x1 = 0;
					drawGrid(div_);
				}
			}
			XFlush(d1);
		}
	}
	exit(EXIT_SUCCESS);
}
